Automatic refresh of Authorization Code Flow Tokens in long-running Applications

MaZderMind opened the pull request:
This PR implements automatic refreshing of Authorization Code Flow Tokens.
Currently an `access_token` requested via the Authorization Code Flow have a lifetime of 3600 seconds (1 hour) after which a new `access_token` needs to be requested using the 'refresh_token`.

The Logic in `util.prompt_for_user_token` looks for a matching set of tokens in the `cache_path` and if none is found, it requests both an `access_token` and a `refresh_token` from Spotify and stores them in the `cache_path`. If one is found, the `access_token` is checked for validity and if it is expired, a new one is requested and updated in the `cache_path`.

This process works fine for short running applications, because the Token is automatically refreshed, whenever `util.prompt_for_user_token` is invoked. For long running applications (with a runtime greater then the remaining lifetime of the `access_token` upon startup) this poses a problem, because the `access_token` is never refreshed, once `util.prompt_for_user_token` has returned.

To refresh the token later, the initialized instance of `oauth2.SpotifyOAuth` is required. This instance is created in `util.prompt_for_user_token` but destroyed when the method returns.

This PR changes the return-value of `util.prompt_for_user_token` to return the initialized `oauth2.SpotifyOAuth` instead of the `access_token`. This instance is accepted as the `auth`-argument to the `spotipy.Spotify` constructor, so code that was just passing the token-string don't have to be changed.

Furthermore `oauth2.SpotifyOAuth` is changed to also cache the `access_token` in-memory, so that accessing it via the oauth2-instance does not have to hit the filesystem.

The method `Spotify._auth_headers` is responsible for adding the Auth-Token to the HTTP Request-Headers. It is changed to actively request a valid `access_token` from the oauth2 instance. The `oauth2.SpotifyOAuth` class is changed to check the validity of the Token before returning and, if required, requesting a new Token.

An Example of a long-running process which before exposed the problem is added. It does not run for multiple hours, refreshing the Token when it has expired before issuing any other request.

This probably fixes #263 and fixes #87
It probably also closes #428 by obsoleting it

MaZderMind commented:
I'll look into the failing tests.

stephanebruckert commented on spotipy/oauth2.py:
```diff
@@ -213,7 +222,7 @@ def parse_response_code(self, url):
         return _make_authorization_headers(self.client_id, self.client_secret)
 
-    def get_access_token(self, code):
+    def request_access_token(self, code):
```
This looks like a public method so if possible I would avoid renaming it

MaZderMind replied:
Yes, you're right about that. I intended to add another `get_access_token` method which would just return the possibly refreshed token, but found that `get_cached_token` is basically doing this exact thing. I removed this change from the commit and force-pushed an update.

MaZderMind commented:
I removed the offending method-name change from the PR and added a Line to the changelog. I also extracted a duplicate code-fragment into a new `_refresh_token_if_expired` method in `oauth2.SpotifyOAuth`

stephanebruckert commented:
Thank you!

ritiek commented:
@MaZderMind Sorry, but this had to be reverted since it seems causing #430. We could re-merge this once the issue is fixed.

ritiek commented on spotipy/client.py:
```diff
@@ -94,7 +94,9 @@ def __init__(self, auth=None, requests_session=True,
-            return {'Authorization': 'Bearer {0}'.format(self._auth)}
+            token_info = self._auth.get_cached_token()
+            access_token = token_info['access_token']
+            return {'Authorization': 'Bearer {0}'.format(access_token)}
```
These changes seem to cause #430. Is there a good reason we need these changes?

stephanebruckert commented on spotipy/client.py:
```diff
@@ -63,7 +63,7 @@ def __init__(self, auth=None, requests_session=True,
         Creates a Spotify API client.
 
-        :param auth: An authorization token (optional)
+        :param auth: An authenticated OAuth2 Instance
```
For backward-compatibility, this parameter should still be "An authorization token"

stephanebruckert commented on spotipy/util.py:
```diff
@@ -103,9 +103,7 @@ def prompt_for_user_token(username, scope=None, client_id=None,
-        return token_info['access_token']
-    else:
-        return None
+    return sp_oauth
```
Also here, to be backward-compatible, we should still return an access token

stephanebruckert commented:
I'm working on this in a new branch https://github.com/plamere/spotipy/tree/change/auto-refresh-token
